### PR TITLE
Make the Linux TLS hostname comparison be case-insensitive

### PR DESCRIPTION
**Pull request by bartonjs**
Solves #30388 in master.

### CODE REVIEW

File: src/System.Security.Cryptography.X509Certificates/tests/HostnameMatchTests.Unix.cs [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+            RunTest(targetName, subjectCN, null, !mixedCase, !wantsWildcard);
+        }
+
+        [Theory]
+        [InlineData("Capitalized.SomeDomain.TLD", false, true)]
+        [InlineData("Capitalized.SomeDomain.TLD", true, true)]
+        [InlineData("Too.Many.SomeDomain.TLD", false, false)]
+        [InlineData("Too.Many.SomeDomain.TLD", true, false)]
+        [InlineData("Now.Lower.SomeDomain.TLD", false, true)]
+        [InlineData("Now.Lower.SomeDomain.TLD", true, true)]
```
Comment (stephentoub) [inline src/System.Security.Cryptography.X509Certificates/tests/HostnameMatchTests.Unix.cs]: The code changed above includes checks for numbers... we should have some tests for numbers included if we don't already.

File: src/System.Security.Cryptography.X509Certificates/tests/HostnameMatchTests.Unix.cs [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+            string targetName,
+            string subjectCN,
+            IList<string> sanDnsNames,
+            bool flattenCase,
+            bool expectedResult)
+        {
+            using (RSA rsa = RSA.Create(TestData.RsaBigExponentParams))
+            {
+                CertificateRequest request = new CertificateRequest(
+                    $"CN={FixCase(subjectCN, flattenCase)}, O=.NET Framework (CoreFX)",
```
Comment (stephentoub) [inline src/System.Security.Cryptography.X509Certificates/tests/HostnameMatchTests.Unix.cs]: FixCase here is just a way of getting more coverage by exercising both the cased and already-equivalent-cased conditions?

File: src/System.Security.Cryptography.X509Certificates/tests/HostnameMatchTests.Unix.cs [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+                {
+                    var builder = new SubjectAlternativeNameBuilder();
+
+                    foreach (string sanDnsName in sanDnsNames)
+                    {
+                        builder.AddDnsName(FixCase(sanDnsName, flattenCase));
+                    }
+
+                    X509Extension extension = builder.Build();
+                    Console.WriteLine(extension.ToString());
```
Comment (stephentoub) [inline src/System.Security.Cryptography.X509Certificates/tests/HostnameMatchTests.Unix.cs]: Can we remove this Console.WriteLine?

File: src/System.Security.Cryptography.X509Certificates/tests/HostnameMatchTests.Unix.cs [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+            int value = CheckX509Hostname(cert.Handle, targetName, targetName.Length);
+            GC.KeepAlive(cert);
+            Assert.InRange(value, 0, 1);
+            return value != 0;
+        }
+
+        [DllImport(Interop.Libraries.CryptoNative, EntryPoint = "CryptoNative_CheckX509Hostname")]
+        private static extern int CheckX509Hostname(IntPtr x509, string hostname, int cchHostname);
+    }
+}
```
Comment (stephentoub) [inline src/System.Security.Cryptography.X509Certificates/tests/HostnameMatchTests.Unix.cs]: Is there a test we could add at the System.Net.Http level to ensure everything works correctly here above the X509Certificate stack, too?

File: src/System.Security.Cryptography.X509Certificates/tests/HostnameMatchTests.Unix.cs (hunk 2 above)
Comment (bartonjs) [reply src/System.Security.Cryptography.X509Certificates/tests/HostnameMatchTests.Unix.cs]: Yep, if case normalization changes the answer something regressed.

File: src/System.Security.Cryptography.X509Certificates/tests/HostnameMatchTests.Unix.cs (hunk 3 above)
Comment (bartonjs) [reply src/System.Security.Cryptography.X509Certificates/tests/HostnameMatchTests.Unix.cs]: Oops :)

File: src/System.Security.Cryptography.X509Certificates/tests/HostnameMatchTests.Unix.cs (hunk 4 above)
Comment (bartonjs) [reply src/System.Security.Cryptography.X509Certificates/tests/HostnameMatchTests.Unix.cs]: Yeah, I started by doing that, with a weirdly-cased localhost cert.  But I didn't know enough about the SslStream test mechanisms to know really how to write it.

File: src/Native/Unix/System.Security.Cryptography.Native/openssl.c [hunk 5]
```diff
@@ -736,7 +736,8 @@ static int CheckX509HostnameMatch(ASN1_STRING* candidate, const char* hostname,
         {
             char c = candidateStr[i];
 
-            if ((c < 'a' || c > 'z') && (c < '0' || c > '9') && (c != '.') && (c != '-') && (c != '*' || i != 0))
+            if ((c < 'A' || c > 'Z') && (c < 'a' || c > 'z') && (c < '0' || c > '9') && (c != '.') && (c != '-') &&
+                (c != '*' || i != 0))
```
Comment (danmosemsft) [inline src/Native/Unix/System.Security.Cryptography.Native/openssl.c]: FWIW I always find AND (NOT x OR NOT y) harder to read than AND NOT (x AND y) 

Given `(c != '*' || i != 0)` I need to mentally decode whereas `!(c == '*' && i == 0)` naturally reads as "no asterisk at the front".

**Pull request merged**